flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/TypeInference.java

```diff
@@ -46,6 +46,8 @@ public final class TypeInference {
 
     private final @Nullable List<String> namedArguments;
 
+    private final @Nullable List<Boolean> optionalArguments;
```

Comment:
I am worried about the current design of the Function Overload and Optional Parameter.

The current implementation suggests that a UDF can only use either function overloading or optional parameters, not both.  But in some situation, we may need both abilities.

```
class ToBytes extends ScalarFunction {

    bytes[] eval(String input, String format ="gzip") {}

    bytes[] eval(int input, String format = "gzip") {}
}
```

Actually, I think the argument optionals and names(alias) should belong to the `InputTypeStrategy`.  It's better we should open another FLIP to discuss the API design, cc @xuyangzhong @lincoln-lil